File: apps/web/src/api/diagnostics.ts
// Fetchers for the gateway's /v1/diagnostics/* endpoints. Returns
// already-shaped data (no envelope unwrap). Auth is handled by the
// shared apiFetch helper — copied locally because the existing one in
// client.ts doesn't export apiFetch and is currently focused on the
// Subsonic envelope. We keep this module standalone so a future
// extraction is mechanical.

import { refreshTokens } from "../auth/oauth";
import { clearTokens, readTokens } from "../auth/tokens";

class AuthError extends Error {}

async function apiFetch(path: string): Promise<Response> {
  const tokens = readTokens();
  if (!tokens) throw new AuthError("not signed in");

  const doFetch = (token: string) =>
    fetch(path, { headers: { Authorization: `Bearer ${token}` } });

  let res = await doFetch(tokens.accessToken);
  if (res.status === 401) {
    try {
      await refreshTokens(tokens.refreshToken);
    } catch {
      clearTokens();
      throw new AuthError("session expired");
    }
    const refreshed = readTokens();
    if (!refreshed) throw new AuthError("session expired");
    res = await doFetch(refreshed.accessToken);
    if (res.status === 401) {
      clearTokens();
      throw new AuthError("session expired");
    }
  }
  return res;
}

async function getJson<T>(path: string): Promise<T> {
  const res = await apiFetch(path);
  if (!res.ok) throw new Error(`HTTP ${res.status} from ${path}`);
  return (await res.json()) as T;
}

async function postJson<T>(path: string): Promise<T> {
  const tokens = readTokens();
  if (!tokens) throw new AuthError("not signed in");
  const res = await fetch(path, {
    method: "POST",
    headers: { Authorization: `Bearer ${tokens.accessToken}` },
  });
  if (!res.ok) throw new Error(`HTTP ${res.status} from ${path}`);
  return (await res.json()) as T;
}

export interface InvalidateCacheResponse {
  removed: number;
}

/** Flush the gateway's L2 browse cache. Returns the row count removed. */
export function invalidateBrowseCache(): Promise<InvalidateCacheResponse> {
  return postJson<InvalidateCacheResponse>("/v1/admin/cache/invalidate");
}

export interface TraceEntry {
  trace_id: string;
  span_id: number;
  parent_span_id: number | null;
  name: string;
  target: string;
  start_ms: number;
  end_ms: number;
  duration_ms: number;
  // Free-form: the gateway parses fields_json server-side, falling back
  // to {"_raw": "..."} on parse failure.
  fields: Record<string, unknown>;
}

export interface TracesResponse {
  traces: TraceEntry[];
}

export interface HistogramBucket {
  name: string;
  count: number;
  min_ms: number;
  max_ms: number;
  p50_ms: number;
  p95_ms: number;
  p99_ms: number;
  mean_ms: number;
}

export interface HistogramResponse {
  buckets: HistogramBucket[];
}

export interface QueueDepthResponse {
  model_version: string;
  not_started: number;
  in_progress: number;
  done: number;
  failed: number;
}

export function fetchTraces(opts: {
  limit?: number;
  name?: string;
  sinceMs?: number;
}): Promise<TracesResponse> {
  const qs = new URLSearchParams();
  if (opts.limit !== undefined) qs.set("limit", String(opts.limit));
  if (opts.name) qs.set("name", opts.name);
  if (opts.sinceMs !== undefined) qs.set("since_ms", String(opts.sinceMs));
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<TracesResponse>(`/v1/diagnostics/traces${suffix}`);
}

export function fetchHistogram(opts: { sinceMs?: number }): Promise<HistogramResponse> {
  const qs = new URLSearchParams();
  if (opts.sinceMs !== undefined) qs.set("since_ms", String(opts.sinceMs));
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<HistogramResponse>(`/v1/diagnostics/histogram${suffix}`);
}

// --- span_series ----------------------------------------------------------
//
// Time-series of `(end_ms, duration_ms)` for a single span name. Powers
// the per-row plot when a histogram row is expanded on /diagnostics/tracing.

export interface SpanSeriesPoint {
  end_ms: number;
  duration_ms: number;
}

export interface SpanSeriesResponse {
  name: string;
  points: SpanSeriesPoint[];
}

export function fetchSpanSeries(opts: {
  name: string;
  sinceMs?: number;
  limit?: number;
}): Promise<SpanSeriesResponse> {
  const qs = new URLSearchParams();
  qs.set("name", opts.name);
  if (opts.sinceMs !== undefined) qs.set("since_ms", String(opts.sinceMs));
  if (opts.limit !== undefined) qs.set("limit", String(opts.limit));
  return getJson<SpanSeriesResponse>(`/v1/diagnostics/span_series?${qs.toString()}`);
}

// --- span_children: parent → direct-child wall-time breakdown ------------

export interface SpanChildAgg {
  name: string;
  count: number;
  sum_ms: number;
  mean_ms: number;
}

export interface SpanChildrenResponse {
  parent_name: string;
  parent_count: number;
  parent_sum_ms: number;
  children: SpanChildAgg[];
}

export function fetchSpanChildren(opts: {
  name: string;
  sinceMs?: number;
}): Promise<SpanChildrenResponse> {
  const qs = new URLSearchParams();
  qs.set("name", opts.name);
  if (opts.sinceMs !== undefined) qs.set("since_ms", String(opts.sinceMs));
  return getJson<SpanChildrenResponse>(`/v1/diagnostics/span_children?${qs.toString()}`);
}

export function fetchQueueDepth(opts: { modelVersion?: string }): Promise<QueueDepthResponse> {
  const qs = new URLSearchParams();
  if (opts.modelVersion) qs.set("model_version", opts.modelVersion);
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<QueueDepthResponse>(`/v1/diagnostics/queue_depth${suffix}`);
}

// --- client events --------------------------------------------------------

export interface ClientEventEntry {
  received_ms: number;
  occurred_ms: number;
  session_id: string;
  name: string;
  value_ms: number | null;
  rating: "good" | "needs-improvement" | "poor" | null;
  page_path: string;
  user_agent: string | null;
  fields: Record<string, unknown>;
}

export interface ClientEventsResponse {
  events: ClientEventEntry[];
}

export function fetchClientEvents(opts: {
  limit?: number;
  name?: string;
}): Promise<ClientEventsResponse> {
  const qs = new URLSearchParams();
  if (opts.limit !== undefined) qs.set("limit", String(opts.limit));
  if (opts.name) qs.set("name", opts.name);
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<ClientEventsResponse>(`/v1/diagnostics/client_events${suffix}`);
}

// --- recently played ------------------------------------------------------

export interface RecentlyPlayedEntry {
  track_id: string;
  /** Client-supplied scrobble timestamp (unix-ms). */
  occurred_at_ms: number;
  /** Gateway-stamped persist time (unix-ms). */
  received_at_ms: number;
  /** All metadata fields are nullable: a track that has been scrobbled
   *  but not yet ingested by the recommender (so it's missing from the
   *  `track_metadata` cache) will return null for every field below.
   *  The UI falls back to the raw `track_id`. */
  title: string | null;
  artist: string | null;
  artist_id: string | null;
  album: string | null;
  album_id: string | null;
  year: number | null;
}

export interface RecentlyPlayedResponse {
  events: RecentlyPlayedEntry[];
}

export function fetchRecentlyPlayed(opts: {
  limit?: number;
  sinceMs?: number;
}): Promise<RecentlyPlayedResponse> {
  const qs = new URLSearchParams();
  if (opts.limit !== undefined) qs.set("limit", String(opts.limit));
  if (opts.sinceMs !== undefined) qs.set("since_ms", String(opts.sinceMs));
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<RecentlyPlayedResponse>(`/v1/diagnostics/recently_played${suffix}`);
}

// --- recommender aggregates ----------------------------------------------
//
// The four /v1/diagnostics/recommend/* aggregations all key off the same
// recommend.from_any / recommend.from_seeds spans, but slice differently:
// fill-ratio histogram, shortfall reason counts, admitted-similarity
// quantiles, and a recommend-frequency leaderboard.

export interface QueueFillBucket {
  label: string;
  count: number;
}

export interface QueueFillResponse {
  total: number;
  buckets: QueueFillBucket[];
}

export interface ShortfallResponse {
  total: number;
  /** Map of `shortfall_reason` → count. Pre-R1 spans land under `unknown`. */
  counts: Record<string, number>;
}

export interface SimilarityResponse {
  count: number;
  p50: number;
  p90: number;
  p95: number;
  p99: number;
  min: number;
  max: number;
  mean: number;
}

export interface TopResultItem {
  track_id: string;
  count: number;
  title: string | null;
  artist: string | null;
  album: string | null;
}

export interface TopResultsResponse {
  items: TopResultItem[];
}

export function fetchRecommendQueueFill(opts: {
  sinceMs?: number;
}): Promise<QueueFillResponse> {
  const qs = new URLSearchParams();
  if (opts.sinceMs !== undefined) qs.set("since_ms", String(opts.sinceMs));
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<QueueFillResponse>(`/v1/diagnostics/recommend/queue_fill${suffix}`);
}

export function fetchRecommendShortfall(opts: {
  sinceMs?: number;
}): Promise<ShortfallResponse> {
  const qs = new URLSearchParams();
  if (opts.sinceMs !== undefined) qs.set("since_ms", String(opts.sinceMs));
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<ShortfallResponse>(`/v1/diagnostics/recommend/shortfall${suffix}`);
}

export function fetchRecommendSimilarity(opts: {
  sinceMs?: number;
}): Promise<SimilarityResponse> {
  const qs = new URLSearchParams();
  if (opts.sinceMs !== undefined) qs.set("since_ms", String(opts.sinceMs));
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<SimilarityResponse>(`/v1/diagnostics/recommend/similarity${suffix}`);
}

export function fetchRecommendTopResults(opts: {
  limit?: number;
  sinceMs?: number;
}): Promise<TopResultsResponse> {
  const qs = new URLSearchParams();
  if (opts.limit !== undefined) qs.set("limit", String(opts.limit));
  if (opts.sinceMs !== undefined) qs.set("since_ms", String(opts.sinceMs));
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<TopResultsResponse>(
    `/v1/diagnostics/recommend/top_results${suffix}`
  );
}

// --- latent space (UMAP scatter) -----------------------------------------

export interface LatentSpacePoint {
  track_id: string;
  x: number;
  y: number;
  title: string | null;
  artist: string | null;
  album: string | null;
  /** Reported genre from Subsonic metadata; null when the track has no
   *  genre tag, or no metadata row in the gateway cache yet. The scatter
   *  uses this to colour clusters as a validation of the embedding. */
  genre: string | null;
  /** First four PCA components on the original 512-D CLAP space,
   *  computed by the reducer alongside (x, y). Null per-component on
   *  projections that predate migration 0009 or for components past
   *  the dataset's natural rank. Used by the "colour by → PCn" mode. */
  pc1: number | null;
  pc2: number | null;
  pc3: number | null;
  pc4: number | null;
  /** Third UMAP axis from an `n_components=3` reducer run (migration
   *  0010). Null on 2-D projections. Drives the vertical spatial axis
   *  in the 3-D scene; not exposed as a colour channel since spatial
   *  position already encodes it. */
  z: number | null;
}

export interface LatentSpaceVersionEntry {
  proj_version: string;
  point_count: number;
  created_at_ms: number;
}

export interface LatentSpaceResponse {
  model_version: string;
  /** null when no projection has been written yet for `model_version`. */
  proj_version: string | null;
  points: LatentSpacePoint[];
  versions: LatentSpaceVersionEntry[];
}

export function fetchRecommendLatentSpace(opts: {
  projVersion?: string;
  modelVersion?: string;
  /** Layout preference, server-resolved to the newest matching
   *  projection. Ignored when `projVersion` is explicitly set. */
  prefer?: "2d" | "3d";
}): Promise<LatentSpaceResponse> {
  const qs = new URLSearchParams();
  if (opts.projVersion) qs.set("proj_version", opts.projVersion);
  if (opts.modelVersion) qs.set("model_version", opts.modelVersion);
  if (opts.prefer) qs.set("prefer", opts.prefer);
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<LatentSpaceResponse>(
    `/v1/diagnostics/recommend/latent_space${suffix}`
  );
}

// --- latent neighbours (hover overlay) -----------------------------------
//
// Returns the k nearest neighbours of a seed track in the original
// CLAP embedding space — the ground-truth distances UMAP can't
// preserve in 2D. The web client fetches this on hover (debounced)
// and uses it to draw ring + connecting-line overlays on the scatter.

export interface LatentNeighbourEntry {
  track_id: string;
  /** Cosine distance in CLAP space (`1 - cosine_similarity`). Range
   *  `[0, 2]`; 0 = identical direction, 1 = orthogonal. */
  cosine_distance: number;
}

export interface LatentNeighboursResponse {
  /** Echoed back so a stale fetch can be detected by the caller. */
  track_id: string;
  /** Ascending by `cosine_distance`. Seed is filtered out server-side. */
  neighbours: LatentNeighbourEntry[];
}

export function fetchRecommendLatentNeighbours(opts: {
  trackId: string;
  k?: number;
}): Promise<LatentNeighboursResponse> {
  const qs = new URLSearchParams();
  qs.set("track_id", opts.trackId);
  if (opts.k !== undefined) qs.set("k", String(opts.k));
  return getJson<LatentNeighboursResponse>(
    `/v1/diagnostics/recommend/latent_neighbours?${qs.toString()}`,
  );
}

// --- recommend sessions --------------------------------------------------

export interface SessionEvent {
  track_id: string;
  event_type: string;
  occurred_at_ms: number;
}

export interface SessionSegment {
  /** Cosine distance between this event's track and the next event's
   *  track in the recommender's CLAP embedding space. `null` when one
   *  of the tracks lacks a `done` embedding under the active model. */
  cosine_distance: number | null;
}

export interface SessionItem {
  session_id: string;
  anchor_track_id: string;
  items_count: number;
  started_ms: number;
  /** null while the session is still active. */
  ended_ms: number | null;
  event_count: number;
  /** Present only when `includeEvents=true`. Oldest-first. */
  events?: SessionEvent[];
  /** Present only when `includeEvents=true`. Length = events.length - 1. */
  segments?: SessionSegment[];
}

export interface SessionsListResponse {
  items: SessionItem[];
}

export function fetchRecommendSessions(opts: {
  limit?: number;
  includeEvents?: boolean;
  modelVersion?: string;
}): Promise<SessionsListResponse> {
  const qs = new URLSearchParams();
  if (opts.limit !== undefined) qs.set("limit", String(opts.limit));
  if (opts.includeEvents) qs.set("include_events", "1");
  if (opts.modelVersion) qs.set("model_version", opts.modelVersion);
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return getJson<SessionsListResponse>(
    `/v1/diagnostics/recommend/sessions${suffix}`
  );
}

File: crates/music-cli/src/api/diagnostics.rs
//! Read-only fetchers for the gateway's `/v1/diagnostics/*` endpoints, powering
//! the TUI's admin-only Diagnostics section (parity plan Phase 9). Every call
//! is a plain bearer-authed GET; the gateway 403s non-admins (the section is
//! hidden for them, but the server is the real gate — decision D3), surfaced as
//! [`ApiError::Forbidden`]. Field names mirror the gateway's serde structs in
//! `music-gateway/src/diagnostics/handlers.rs` verbatim (no `rename`s there).

use anyhow::{Context, anyhow};
use serde::Deserialize;

use super::ApiError;
use crate::config::Config;
use crate::gateway::{endpoint, http_client, require_gateway};

// ── ingest ────────────────────────────────────────────────────────────────

/// `queue_depth`: embedding-ingest queue counts for one model version.
#[derive(Debug, Clone, Deserialize)]
pub struct QueueDepth {
    pub model_version: String,
    pub not_started: u64,
    pub in_progress: u64,
    pub done: u64,
    pub failed: u64,
}

// ── tracing ───────────────────────────────────────────────────────────────

/// One closed span from `traces`. `fields` is free-form JSON (the gateway
/// stores every span's structured fields; a parse failure arrives as
/// `{"_raw": "…"}`).
#[derive(Debug, Clone, Deserialize)]
pub struct TraceEntry {
    pub trace_id: String,
    pub span_id: i64,
    pub parent_span_id: Option<i64>,
    pub name: String,
    pub target: String,
    pub start_ms: i64,
    pub end_ms: i64,
    pub duration_ms: i64,
    #[serde(default)]
    pub fields: serde_json::Value,
}

#[derive(Debug, Clone, Deserialize)]
struct TracesResponse {
    traces: Vec<TraceEntry>,
}

/// One per-span-name duration bucket from `histogram` (quantiles are computed
/// server-side via nearest-rank).
#[derive(Debug, Clone, Deserialize)]
pub struct HistogramBucket {
    pub name: String,
    pub count: usize,
    pub min_ms: i64,
    pub max_ms: i64,
    pub p50_ms: i64,
    pub p95_ms: i64,
    pub p99_ms: i64,
    pub mean_ms: f64,
}

#[derive(Debug, Clone, Deserialize)]
struct HistogramResponse {
    buckets: Vec<HistogramBucket>,
}

// ── client events (RUM) ────────────────────────────────────────────────────

/// One browser RUM event from `client_events`. Two timestamps: `occurred_ms`
/// (client clock) and `received_ms` (gateway-stamped).
#[derive(Debug, Clone, Deserialize)]
pub struct ClientEvent {
    pub received_ms: i64,
    pub occurred_ms: i64,
    pub session_id: String,
    pub name: String,
    pub value_ms: Option<f64>,
    pub rating: Option<String>,
    pub page_path: String,
    pub user_agent: Option<String>,
    #[serde(default)]
    pub fields: serde_json::Value,
}

#[derive(Debug, Clone, Deserialize)]
struct ClientEventsResponse {
    events: Vec<ClientEvent>,
}

// ── listening (recently played) ────────────────────────────────────────────

/// One `recently_played` row: a play event with its (best-effort) hydrated
/// track metadata.
#[derive(Debug, Clone, Deserialize)]
pub struct RecentlyPlayed {
    pub track_id: String,
    pub occurred_at_ms: i64,
    pub received_at_ms: i64,
    pub title: Option<String>,
    pub artist: Option<String>,
    pub artist_id: Option<String>,
    pub album: Option<String>,
    pub album_id: Option<String>,
    pub year: Option<i32>,
}

#[derive(Debug, Clone, Deserialize)]
struct RecentlyPlayedResponse {
    events: Vec<RecentlyPlayed>,
}

// ── recommender inspectors ─────────────────────────────────────────────────

/// One bar of the `recommend/queue_fill` histogram (fixed fill-ratio labels).
#[derive(Debug, Clone, Deserialize)]
pub struct FillBucket {
    pub label: String,
    pub count: u64,
}

/// `recommend/queue_fill`: how full autoplay refills came back, bucketed.
#[derive(Debug, Clone, Deserialize)]
pub struct QueueFill {
    pub total: u64,
    pub buckets: Vec<FillBucket>,
}

/// `recommend/shortfall`: why refills under-delivered (reason → count).
#[derive(Debug, Clone, Deserialize)]
pub struct Shortfall {
    pub total: u64,
    pub counts: std::collections::BTreeMap<String, u64>,
}

/// `recommend/similarity`: distribution of served-candidate cosine scores.
#[derive(Debug, Clone, Deserialize)]
pub struct Similarity {
    pub count: u64,
    pub p50: f64,
    pub p90: f64,
    pub p95: f64,
    pub p99: f64,
    pub min: f64,
    pub max: f64,
    pub mean: f64,
}

/// One `recommend/top_results` leaderboard row: a track and how often it was
/// served as a recommendation in the window.
#[derive(Debug, Clone, Deserialize)]
pub struct TopResultItem {
    pub track_id: String,
    pub count: u64,
    pub title: Option<String>,
    pub artist: Option<String>,
    pub album: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
struct TopResultsResponse {
    items: Vec<TopResultItem>,
}

/// The Recommender sub-tab's four panels, fetched together for one window.
#[derive(Debug, Clone)]
pub struct RecommenderPanels {
    pub queue_fill: QueueFill,
    pub shortfall: Shortfall,
    pub similarity: Similarity,
    pub top_results: Vec<TopResultItem>,
}

// ── latent space ───────────────────────────────────────────────────────────

/// One projected point from `recommend/latent_space` (2-D; the `pc*`/`z`
/// fields the gateway also returns are unused here).
#[derive(Debug, Clone, Deserialize)]
pub struct LatentPoint {
    pub track_id: String,
    pub x: f64,
    pub y: f64,
    pub title: Option<String>,
    pub artist: Option<String>,
    pub album: Option<String>,
    pub genre: Option<String>,
}

/// `recommend/latent_space`: the whole 2-D scatter for one projection version.
/// `proj_version` is `None` when the gateway has no projection yet (an empty
/// scatter, not an error).
#[derive(Debug, Clone, Deserialize)]
pub struct LatentSpace {
    pub model_version: String,
    pub proj_version: Option<String>,
    pub points: Vec<LatentPoint>,
}

/// One `recommend/latent_neighbours` entry: a track near the seed in the raw
/// embedding space (cosine distance, smaller = closer).
#[derive(Debug, Clone, Deserialize)]
pub struct LatentNeighbour {
    pub track_id: String,
    pub cosine_distance: f64,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub artist: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
struct LatentNeighboursResponse {
    neighbours: Vec<LatentNeighbour>,
}

// ── HTTP plumbing ──────────────────────────────────────────────────────────

/// A `since_ms` query fragment: one pair, or empty for "all time".
fn since_query(since_ms: Option<i64>) -> Vec<(&'static str, String)> {
    since_ms.map_or_else(Vec::new, |ms| vec![("since_ms", ms.to_string())])
}

/// Shared bearer-authed GET + JSON parse, with 403 mapped to
/// [`ApiError::Forbidden`] (non-admin) so the caller renders an honest notice.
async fn diag_get<T: serde::de::DeserializeOwned>(
    config: &Config,
    path: &str,
    query: &[(&str, String)],
) -> Result<T, ApiError> {
    let gw = require_gateway(config)?;
    let token = crate::auth::resolve_bearer(config, gw).await?;
    let url = endpoint(gw, path);
    let resp = http_client(gw)?
        .get(&url)
        .query(query)
        .bearer_auth(&token)
        .send()
        .await
        .context("requesting diagnostics")?;
    if resp.status().as_u16() == 403 {
        return Err(ApiError::Forbidden);
    }
    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        // A latent seed with no stored vector (or no projection) 404s — not a
        // hard error; surface it the same as a warming recommender so the UI
        // shows a friendly notice.
        return Err(ApiError::RecommenderUnavailable);
    }
    if !resp.status().is_success() {
        let status = resp.status();
        let text = resp.text().await.unwrap_or_default();
        return Err(anyhow!("diagnostics request failed ({status}): {text}").into());
    }
    resp.json()
        .await
        .context("parsing diagnostics response")
        .map_err(Into::into)
}

/// `GET /v1/diagnostics/queue_depth` — ingest queue counts (default model).
pub async fn queue_depth(config: &Config) -> Result<QueueDepth, ApiError> {
    diag_get(config, "/v1/diagnostics/queue_depth", &[]).await
}

/// `GET /v1/diagnostics/traces?limit=` — recent closed spans, newest first.
pub async fn traces(config: &Config, limit: usize) -> Result<Vec<TraceEntry>, ApiError> {
    let r: TracesResponse = diag_get(
        config,
        "/v1/diagnostics/traces",
        &[("limit", limit.to_string())],
    )
    .await?;
    Ok(r.traces)
}

/// `GET /v1/diagnostics/histogram?since_ms=` — per-span-name duration buckets.
pub async fn histogram(
    config: &Config,
    since_ms: Option<i64>,
) -> Result<Vec<HistogramBucket>, ApiError> {
    let r: HistogramResponse =
        diag_get(config, "/v1/diagnostics/histogram", &since_query(since_ms)).await?;
    Ok(r.buckets)
}

/// `GET /v1/diagnostics/client_events?limit=` — recent browser RUM events.
pub async fn client_events(config: &Config, limit: usize) -> Result<Vec<ClientEvent>, ApiError> {
    let r: ClientEventsResponse = diag_get(
        config,
        "/v1/diagnostics/client_events",
        &[("limit", limit.to_string())],
    )
    .await?;
    Ok(r.events)
}

/// `GET /v1/diagnostics/recently_played?limit=` — recent play events.
pub async fn recently_played(
    config: &Config,
    limit: u32,
) -> Result<Vec<RecentlyPlayed>, ApiError> {
    let r: RecentlyPlayedResponse = diag_get(
        config,
        "/v1/diagnostics/recently_played",
        &[("limit", limit.to_string())],
    )
    .await?;
    Ok(r.events)
}

async fn queue_fill(config: &Config, since_ms: Option<i64>) -> Result<QueueFill, ApiError> {
    diag_get(
        config,
        "/v1/diagnostics/recommend/queue_fill",
        &since_query(since_ms),
    )
    .await
}

async fn shortfall(config: &Config, since_ms: Option<i64>) -> Result<Shortfall, ApiError> {
    diag_get(
        config,
        "/v1/diagnostics/recommend/shortfall",
        &since_query(since_ms),
    )
    .await
}

async fn similarity(config: &Config, since_ms: Option<i64>) -> Result<Similarity, ApiError> {
    diag_get(
        config,
        "/v1/diagnostics/recommend/similarity",
        &since_query(since_ms),
    )
    .await
}

async fn top_results(
    config: &Config,
    since_ms: Option<i64>,
    limit: usize,
) -> Result<Vec<TopResultItem>, ApiError> {
    let mut query = since_query(since_ms);
    query.push(("limit", limit.to_string()));
    let r: TopResultsResponse =
        diag_get(config, "/v1/diagnostics/recommend/top_results", &query).await?;
    Ok(r.items)
}

/// The Recommender sub-tab: all four inspectors for one window, fetched
/// concurrently. A single failure fails the whole panel (they share a window
/// and target the same gateway — a failure is systemic, not per-panel).
pub async fn recommender_panels(
    config: &Config,
    since_ms: Option<i64>,
) -> Result<RecommenderPanels, ApiError> {
    let (queue_fill, shortfall, similarity, top_results) = tokio::try_join!(
        queue_fill(config, since_ms),
        shortfall(config, since_ms),
        similarity(config, since_ms),
        top_results(config, since_ms, TOP_RESULTS_LIMIT),
    )?;
    Ok(RecommenderPanels {
        queue_fill,
        shortfall,
        similarity,
        top_results,
    })
}

/// `GET /v1/diagnostics/recommend/latent_space` — the 2-D scatter (newest
/// projection). Empty `points` when no projection exists yet.
pub async fn latent_space(config: &Config) -> Result<LatentSpace, ApiError> {
    diag_get(
        config,
        "/v1/diagnostics/recommend/latent_space",
        &[("prefer", "2d".to_owned())],
    )
    .await
}

/// `GET /v1/diagnostics/recommend/latent_neighbours?track_id=&k=` — the raw
/// nearest neighbours of a seed track. 404 (seed not embedded) →
/// [`ApiError::RecommenderUnavailable`].
pub async fn latent_neighbours(
    config: &Config,
    track_id: &str,
    k: usize,
) -> Result<Vec<LatentNeighbour>, ApiError> {
    let r: LatentNeighboursResponse = diag_get(
        config,
        "/v1/diagnostics/recommend/latent_neighbours",
        &[
            ("track_id", track_id.to_owned()),
            ("k", k.to_string()),
        ],
    )
    .await?;
    Ok(r.neighbours)
}

/// Rows requested for the top-results leaderboard.
const TOP_RESULTS_LIMIT: usize = 20;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn queue_depth_parses() {
        let q: QueueDepth = serde_json::from_str(
            r#"{"model_version":"clamp3","not_started":3,"in_progress":1,"done":7349,"failed":31}"#,
        )
        .unwrap();
        assert_eq!(q.done, 7349);
        assert_eq!(q.failed, 31);
    }

    #[test]
    fn shortfall_counts_is_a_map() {
        let s: Shortfall = serde_json::from_str(
            r#"{"total":5,"counts":{"unindexed_seed":3,"unknown":2}}"#,
        )
        .unwrap();
        assert_eq!(s.total, 5);
        assert_eq!(s.counts.get("unindexed_seed"), Some(&3));
    }

    #[test]
    fn latent_space_tolerates_no_projection() {
        // proj_version null + empty points = "no projection yet", not an error.
        let l: LatentSpace =
            serde_json::from_str(r#"{"model_version":"clamp3","proj_version":null,"points":[]}"#)
                .unwrap();
        assert!(l.proj_version.is_none());
        assert!(l.points.is_empty());
    }

    #[test]
    fn latent_point_ignores_extra_pc_fields() {
        let l: LatentSpace = serde_json::from_str(
            r#"{"model_version":"m","proj_version":"m-2d","points":[
                {"track_id":"t1","x":0.1,"y":-0.2,"title":"A","artist":"B","album":"C",
                 "genre":"Jazz","pc1":1.0,"pc2":2.0,"pc3":null,"pc4":null,"z":null}]}"#,
        )
        .unwrap();
        assert_eq!(l.points.len(), 1);
        assert_eq!(l.points[0].genre.as_deref(), Some("Jazz"));
    }

    #[test]
    fn since_query_omits_when_all_time() {
        assert!(since_query(None).is_empty());
        assert_eq!(since_query(Some(42)), vec![("since_ms", "42".to_owned())]);
    }
}
